glean-core/python/glean/metrics/jwe.py

```diff
@@ -0,0 +1,192 @@
+
+    def set_with_compact_representation(self, value: str):
+        """
+        Set to the specified JWE value.
+
+        Args:
+            value (str): the compact representation of a JWE value.
+        """
+        if self._disabled:
+            return
+
+        @Dispatcher.launch
+        def set():
+            _ffi.lib.glean_jwe_set_with_compact_representation(
+                self._handle, _ffi.ffi_encode_string(value)
+            )
+
+    def set(
+        self, header: str, key: str, init_vector: str, cipher_text: str, auth_tag: str
+    ) -> None:
+        """
+        Build a JWE value from its elements and set to it.
+
+        Args:
+            header (str): A variable-size JWE protected header.
+            key (str): A variable-size encrypted key.
+                This can be an empty octet sequence.
+            init_vector (str): A fixed-size, 96-bit, base64 encoded Jwe initialization vector.
+                If not required by the encryption algorithm, can be an empty octet sequence.
+            cipher_text (str): The variable-size base64 encoded cipher text.
+            auth_tag (str): A fixed-size, 132-bit, base64 encoded authentication tag.
+                Can be an empty octet sequence.
+        """
+        if self._disabled:
+            return
+
+        @Dispatcher.launch
+        def set():
+            _ffi.lib.glean_jwe_set(
+                self._handle,
+                _ffi.ffi_encode_string(header),
+                _ffi.ffi_encode_string(key),
+                _ffi.ffi_encode_string(init_vector),
+                _ffi.ffi_encode_string(cipher_text),
+                _ffi.ffi_encode_string(auth_tag),
+            )
+
+    def test_has_value(self, ping_name: Optional[str] = None) -> bool:
+        """
+        Tests whether a value is stored for the metric for testing purposes
+        only.
+
+        Args:
+            ping_name (str): (default: first value in send_in_pings) The name
+                of the ping to retrieve the metric for.
+
+        Returns:
+            has_value (bool): True if the metric value exists.
+        """
+        if ping_name is None:
+            ping_name = self._send_in_pings[0]
+
+        return bool(
+            _ffi.lib.glean_jwe_test_has_value(
+                self._handle, _ffi.ffi_encode_string(ping_name)
+            )
+        )
+
+    def test_get_value(self, ping_name: Optional[str] = None) -> str:
+        """
+        Returns the stored value for testing purposes only.
+
+        Args:
+            ping_name (str): (default: first value in send_in_pings) The name
+                of the ping to retrieve the metric for.
+
+        Returns:
+            value (int): value of the stored metric.
+        """
+        if ping_name is None:
+            ping_name = self._send_in_pings[0]
+
+        if not self.test_has_value(ping_name):
+            raise ValueError("metric has no value")
+
+        return _ffi.ffi_decode_string(
```

Comment:
Is this is JSON, it's probaly more natural to return the parsed JSON objects rather than a string of JSON. (But I also don't fully understand the use cases...)

Reply:
I agree that it probably makes sense to parse the JSON on the bindings, like I said in a previous response: returning this as a JSON was to follow a guideline from the [design doc](https://docs.google.com/document/d/1nntNIiE6braTGzoKf-lx21OVDd8ssyeIeJu3jnQQfEE/edit). I also am not sure exactly about the use cases, now that I think about it. @Dexterp37 what do you think?

Reply:
In general, we use JSON here and in other types as a convenient way to just pass a string over FFI instead of defining structures. We should probably do this properly, someday, in order to unblock the [Android test speedup as well](https://github.com/mozilla/glean/pull/837#issuecomment-617640739).

Looking back at the spec, we don't ever say that this should return JSON. We say that ` returns an object representing the different parts that make up the JWE token, i.e. header, key, initVector, cipherText, authTag`. A tuple or a struct is fine. 

Good catch @mdboom